diff --git a/src/libsystemd-network/sd-dhcp6-client.c b/src/libsystemd-network/sd-dhcp6-client.c
index 5e71d63de5..e2415a5b47 100644
--- a/src/libsystemd-network/sd-dhcp6-client.c
+++ b/src/libsystemd-network/sd-dhcp6-client.c
@@ -182,12 +182,12 @@ int sd_dhcp6_client_add_vendor_option(sd_dhcp6_client *client, sd_dhcp6_option *
         }
 
         r = ordered_set_ensure_put(&client->vendor_options, &dhcp6_option_hash_ops, v);
-        if (r < 0)
+        if (r <= 0)
                 return r;
 
         sd_dhcp6_option_ref(v);
 
-        return 1;
+        return r;
 }
 
 static int client_ensure_duid(sd_dhcp6_client *client) {
@@ -541,11 +541,11 @@ int sd_dhcp6_client_add_option(sd_dhcp6_client *client, sd_dhcp6_option *v) {
         assert_return(v, -EINVAL);
 
         r = ordered_hashmap_ensure_put(&client->extra_options, &dhcp6_option_hash_ops, UINT_TO_PTR(v->option), v);
-        if (r < 0)
+        if (r <= 0)
                 return r;
 
         sd_dhcp6_option_ref(v);
-        return 0;
+        return r;
 }
 
 static void client_set_state(sd_dhcp6_client *client, DHCP6State state) {
-- 
2.50.1
