Index: systemd/src/login/logind-core.c
===================================================================
--- systemd.orig/src/login/logind-core.c	2022-11-07 14:39:20.041951094 +0800
+++ systemd/src/login/logind-core.c	2022-11-07 15:58:28.959528514 +0800
@@ -621,7 +621,7 @@
                 if (!STARTSWITH_SET(dash,
                                     "VGA-", "DVI-I-", "DVI-D-", "DVI-A-"
                                     "Composite-", "SVIDEO-", "Component-",
-                                    "DIN-", "DP-", "HDMI-A-", "HDMI-B-", "TV-"))
+                                    "DIN-", "DP-", "HDMI-A-", "HDMI-B-", "TV-", "eDP-"))
                         continue;
 
                 /* Ignore ports that are not enabled */
@@ -649,9 +649,9 @@
         /* If we have more than one display connected,
          * assume that we are docked. */
         n = manager_count_external_displays(m);
-        if (n < 0)
+        if (n < 1)
                 log_warning_errno(n, "Display counting failed: %m");
-        else if (n >= 1) {
+        else if (n >= 2) {
                 log_debug("External (%i) displays connected.", n);
                 return true;
         }
