From 23aa22ce36e6cd25ed3b7505ad05620d79ab1811 Mon Sep 17 00:00:00 2001
From: zhouzilong <zhouzilong@uniontech.com>
Date: Sat, 30 Nov 2024 11:28:19 +0800
Subject: [PATCH] change security.selinux2 to security.usec_proc attr

---
 src/core/umac-access.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/core/umac-access.c b/src/core/umac-access.c
index 4ecde3d..8f73f0c 100644
--- a/src/core/umac-access.c
+++ b/src/core/umac-access.c
@@ -30,7 +30,7 @@
 
 static bool initialized = false;
 
-#define DEFAULT_USID_CONTEXT	"root:sysadm_r:sysadm_t:s0"
+#define DEFAULT_USID_CONTEXT	"root:object_r:deepin_usec_t:s0"
 
 /*
    Any time an access gets denied this callback will be called
@@ -139,7 +139,7 @@ static int mac_selinux_getfilecon2(const char *path, char **con) {
                 if (context == NULL)
                         return -ENOMEM;
 
-                r = getxattr(path, "security.selinux2", context, MAX_USID);
+                r = getxattr(path, "security.usec_proc", context, MAX_USID);
                 if (r <= 0) {
                         free(context);
                         context = NULL;
-- 
2.20.1

