diff --git a/tmpfiles.d/x11.conf b/tmpfiles.d/x11.conf
index ef0b11d..acd751c 100644
--- a/tmpfiles.d/x11.conf
+++ b/tmpfiles.d/x11.conf
@@ -9,10 +9,11 @@
 
 # Make sure these are created by default so that nobody else can
 # or empty them at startup
-D! /tmp/.X11-unix 1777 root root 10d
-D! /tmp/.ICE-unix 1777 root root 10d
-D! /tmp/.XIM-unix 1777 root root 10d
-D! /tmp/.font-unix 1777 root root 10d
+D /tmp/.X11-unix 1777 root root 1h
+D /tmp/.ICE-unix 1777 root root 1h
+D /tmp/.XIM-unix 1777 root root 1h
+D /tmp/.font-unix 1777 root root 1h
 
-# Unlink the X11 lock files
+# Unlink the X11 lock files at boot, but exclude them from time-based cleanup later
 r! /tmp/.X[0-9]*-lock
+x /tmp/.X[0-9]*-lock
