Index: systemd/src/basic/path-util.c
===================================================================
--- systemd.orig/src/basic/path-util.c
+++ systemd/src/basic/path-util.c
@@ -793,7 +793,7 @@ static int executable_is_good(const char
         if (r < 0)
                 return r;
 
-        return !PATH_IN_SET(d, "true"
+        return !PATH_IN_SET(d, "true",
                                "/bin/true",
                                "/usr/bin/true",
                                "/dev/null");
