Index: deepin-systemd/src/basic/filesystems-gperf.gperf
===================================================================
--- deepin-systemd.orig/src/basic/filesystems-gperf.gperf	2025-10-29 17:33:51.000000000 +0800
+++ deepin-systemd/src/basic/filesystems-gperf.gperf	2025-10-29 17:34:10.486422106 +0800
@@ -60,6 +60,7 @@
 ext2,            {EXT2_SUPER_MAGIC}
 ext3,            {EXT3_SUPER_MAGIC}
 ext4,            {EXT4_SUPER_MAGIC}
+tpmfs,           {TPMFS_SUPER_MAGIC}
 exfat,           {EXFAT_SUPER_MAGIC}
 f2fs,            {F2FS_SUPER_MAGIC}
 # fuseblk is so closely related to fuse that it shares the same magic
