Index: systemd/src/libsystemd/sd-journal/journal-file.c
===================================================================
--- systemd.orig/src/libsystemd/sd-journal/journal-file.c
+++ systemd/src/libsystemd/sd-journal/journal-file.c
@@ -1894,7 +1894,7 @@ static int journal_file_append_data(
 
         /* ... and link it in. */
         o->data.next_field_offset = fo->field.head_data_offset;
-        fo->field.head_data_offset = le64toh(p);
+        fo->field.head_data_offset = htole64(p);
 
         if (ret_object)
                 *ret_object = o;
