--- a/src/shared/bus-util.c
+++ b/src/shared/bus-util.c
@@ -174,8 +174,9 @@
 
 bool bus_error_is_unknown_service(const sd_bus_error *error) {
         return sd_bus_error_has_names(error,
-                                      SD_BUS_ERROR_SERVICE_UNKNOWN,
                                       SD_BUS_ERROR_NAME_HAS_NO_OWNER,
+                                      SD_BUS_ERROR_SERVICE_UNKNOWN,
+                                      SD_BUS_ERROR_UNKNOWN_OBJECT,
                                       BUS_ERROR_NO_SUCH_UNIT);
 }
 
